### ISSUE TITLE
DbValue.GetDateTime throws a FormatException when converting datetime string to a DateTime value.

### ISSUE THREAD

**Issue by deAtog** [FirebirdSQL/NETProvider#1173]
Under certain circumstances, the call to Convert.ToDateTime in this method will throw a FormatException when attempting to retrieve a DateTime value inserted into a text column:
https://github.com/FirebirdSQL/NETProvider/blob/73b0082bd253efb6e46cdb28b4f558632f431815/src/FirebirdSql.Data.FirebirdClient/Common/DbValue.cs#L246-L255

Steps to reproduce this issue are as follows:
1. Create at table with a text column.
2. Set the current locale to the invariant culture using: `CultureInfo.CurrentCulture = CultureInfo.InvariantCulture;`
3. Insert a DateTime value with the date of "2024-02-29 00:00:00" into the column using a parameterized query with a DateTime parameter.
4. Change the locale to ar-SA using: `CultureInfo.CurrentCulture = CultureInfo.GetCulture("ar-SA");`
5. Select the text value from the table without casting the value in the query to date/time type. 
6. Use FbDataReader.GetDateTime() to retrieve the value and observe the exception.

A format exception is generated because the call to Convert.ToDateTime uses the current culture of the calling application and does not correctly parse the value that was stored in the database.  This method should either attempt a direct cast to a DateTime and throw an InvalidCastException or perform an implict conversion from the string value to a DateTime value using the inverse conversion as performed by the database upon insert. The latter provides consistent insert and select functionality.

It _may_ be sufficient to use the InvariantCulture in the call to Convert.ToDateTime here, but I personally do not know how Firebase converts DateTime values to text. Additional tests may be required for other locale combinations to ensure correct functionality.

**Comment by cincuranet** [comment]
The usage of `CurrentCulture` is certainly wrong. On the other hand this method is not expected to work with strings. If you store `DateTime` in database as a string, you're doing it wrong. Firebird has dedicated datatype(s) for date/time.

**Comment by deAtog** [comment]
Obviously it is not be advisable to store a DateTime in a text column. However there are likely cases where users have done so. When given a DateTime object as a parameter to insert, Firebase silently converts the value to text. The expectation here is that the driver should be able to reverse the conversion from text when asked to do so.

**Comment by BFuerchau** [comment]
You can not convert simply the string to datetime. Look at the stored format (the default is depending from current culture), but in multi computer scenarieos each can have another culture and so a different format.
Change the datetime to string with own format before and use DateTime.TryParse() with the stored format.

Better is, to change the table column to timestamp, which is fully supported.

**Comment by deAtog** [comment]
@BFuerchau At the end of the day a database is nothing more than a storage implementation. How the data is stored is not important so long as what you put in you are able to get back out again. This bug is an example of FirebirdSQL client failing to store and retrieve data in a consistent manner. 

A DateTime parameter is interpreted as a DbDataType.Timestamp as described here:
https://github.com/FirebirdSQL/NETProvider/blob/fcec8034afdea2fc1846bf28a1b4e4f62a4ed166/src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbParameter.cs#L128-L132

The DateTime value is converted to a timestamp using [DbValue.GetDate()](https://github.com/FirebirdSQL/NETProvider/blob/fcec8034afdea2fc1846bf28a1b4e4f62a4ed166/src/FirebirdSql.Data.FirebirdClient/Common/DbValue.cs#L351) and [DbValue.GetTime()](https://github.com/FirebirdSQL/NETProvider/blob/fcec8034afdea2fc1846bf28a1b4e4f62a4ed166/src/FirebirdSql.Data.FirebirdClient/Common/DbValue.cs#L360) and then sent to the server. The server converts the timestamp it receives to text using the [datetime_to_text](
https://github.com/FirebirdSQL/firebird/blob/4dd80eda5d4bf1f41c15fa4795567896b06da39e/src/common/cvt.cpp#L2363) function.

This conversion is straight forward. Given a dbtype_timestamp value, it decodes it into a [struct tm](https://cplusplus.com/reference/ctime/tm/). It then forms a string in the form of "yyyy-mm-dd hh:mm:ss.tttt" or "dd-MMM-yyyy hh:mm:ss.TTTT" where TTTT is optionally provided depending on database version. More importantly, the conversion to string **ignores the locale** on the server and the client. As such, two clients with different locales and the same DateTime value will result in the same text value stored in the database. Using Convert.ToDateTime with the current date/time format of the locale of the local machine is incorrect. The FirebirdSQL client should only attempt to parse the date time according to the forms used by the server.

**Comment by cincuranet** [comment]
If you have date/time stored in a database (or on the client side) as a string, you're asking for trouble. It is dangerous and wrong on so many levels (and yes, the `CurrentCulture` makes it even harder to round trip).

I would personally like to remove the support to convert `string` to `DateTime` from the code, but that would be huge breaking change for no clear win. In regular usage it's basically unused code path anyway.

I'm going to close this issue now as I don't see value changing anything here.

**Comment by deAtog** [comment]
@cincuranet At minimum, the use of `CurrentCulture` here should be removed in favor of using the `InvariantCulture`. **Using `CurrentCulture` results in inconsistent behavior**. An application deployed on one system may work perfectly fine while fail on another. Any opinion about how you think users should store their data is irrelevant. If the FirebirdSQL database did not want to allow users to store a Timestamp in a character field, then they would have explicitly forbid it and thrown an error when a conversion was needed.  The client driver must not throw an exception on one system vs another given the exact same data simply because one client uses a different culture or locale than the other. The value gained by this change is a method that works reliably and consistently.

For reference, this bug was discovered while I was working on changes for NHibernate, a common database abstraction layer similar to Entity Framework. NHibernate currently accesses data from a DataReader using the `Item[int]` indexer. The indexer returns an object which requires a cast to the requested type. Any cast here is problematic because there are many cases where there is not enough information available about the data to perform the cast correctly. For instance, SQLite does not have a DateTime type and requires storing the value in a TEXT, REAL, or INTEGER field where the format is dependent on connection specific settings. Accessing the indexer in these cases returns a string, double, or integer value which must be converted to a DateTime. The consumer of the data should not be responsible for determining how to convert the data in those fields back to the original type. That is purpose of the `DataReader.Get*` methods of the DataReader. These methods provide a **consistent** way to request a specific data type from the result set according to database specific rules. If the database performs a conversion on the data given, the DataReader **must** perform the appropriate conversion back to the original type. That conversion should work on all machines regardless of the culture or locale. As such, NHibernate should call the type specific methods of the DataReader to obtain values instead of using the indexer. However, the GetDateTime() implementation for the FirebirdSQL client driver was found to be problematic in the aforementioned case because it does not produce a consistent result.

As demonstrated, this issue is not inherently specific to FirebirdSQL. Many databases silently convert data upon insert to the destination field type according to a particular set of rules. Those rules vary from database to database. Some rely upon connection string parameters, while others rely on schema properties or server variables. The Get* methods of the DataReader were designed to read data from a result set in a database agnostic way and to perform any conversions according to the database specific rules. The FirebirdSQL client driver is not adhering to the rules of the FirebirdSQL database in this case. The expected result here should be the same as if the query had performed an explicit cast from the character field to a timestamp type.

**Comment by BFuerchau** [comment]
Yes, this is known practice, if a type isn't specified and values are stored in text columns.
In this case, the database manager ist not the responsable to convert data!
I'm programming databases now since over 50 years and it's always practice to convert the data by myself to put in the database or retrieve from the database.
The database manager doesn't know from wich client data comes and to which client data must be sent.
If you take a timestamp with a current datetime object on any client, normally you get a value from the local timezone.
So if you send this value to the database, you have to convert the value to an apropriate value which is compareable to each value at any time and any location in the world. For this exists, since my early years in programming since 1975, already the UTC-Time.
So put the right value in the datebase you can do all, what you want.
If a database doesn't support a specific value so you must store the value in text or binary columns. Than the database can only transport the _value as is_ from and to database.
If your runtime of any language does also not support the correct type, so my be, you choose the wrong software.
I know also, that SQLite stores in the past only text columns. So SQLite was neven an option to use it as a datastore.
If you use JSON/XML for data exchange, is it yours to choose the correct format, the correct codepage (ANSI, UTF8, ...) so the data will by used from anyone everywhere. For this we have invariant culture to do so.
And also, if frameworks like EF doesn't work correct with textcolumns, if you store other values than texts, than not the framework must have an eye on this.
And the Getter of any type must match the type, which is stored in the database.
If i use a GetInt() from Text, the value is tried to convert and i get an error if the value isn't pure numeric.
If i use a GetDouble() from text, the decimal point can be comma or dot and the grouping character vice versa. So which value should come back? Is the value strored as 1,000,000 or 1.000.000 or 1e+6?

It is not a problem of the driver or database, it is a problem of the developer how the software has to be designed.

**Comment by deAtog** [comment]
The problem isn't whether or not frameworks like Entiy Framework or NHibernate work with text columns correctly. They do what they are told to do. If a developer tells Entity Framework or NHibernate that a column contains a DateTime value these frameworks generate a parameterized statement with a DateTime object that is passed to the client driver and sent to the database. When selecting the value from the database, these frameworks expect to be able to get a DateTime value back that was successfully insert into the database. The problem here is that the `FbDataReader.GetDateTime` method only works correctly if the date time format of `CurrentCulture` matches that of the en-US culture or the `InvariantCulture`. You are arguing over a single line change that is the difference between the client driver doing the right thing vs the inconsistent behavior it has now when it unexpectedly throws an exception.

> If i use a GetInt() from Text, the value is tried to convert and i get an error if the value isn't pure numeric.

If an integer value is inserted as an int and then converted to a text value by the database, the client driver should **never** have a problem converting the textual value back to an int. There is **never** a case where inserting an integer via a parameterized query would ever result in a value that could not be parsed back as an integer by the client according to the database specific cast performed. The [integer_to_text](https://github.com/FirebirdSQL/firebird/blob/4dd80eda5d4bf1f41c15fa4795567896b06da39e/src/common/cvt.cpp#L523) function defines how an integer value is converted to text.

> If i use a GetDouble() from text, the decimal point can be comma or dot and the grouping character vice versa. So which value should come back? Is the value strored as 1,000,000 or 1.000.000 or 1e+6?

This is no different than the above. The [float_to_text](https://github.com/FirebirdSQL/firebird/blob/4dd80eda5d4bf1f41c15fa4795567896b06da39e/src/common/cvt.cpp#L323) function defines how a floating point value is converted to text. This function always uses the printf %f and %g specifiers. The format of these specifiers is well defined and always use a dot as the decimal separator when appropriate. These format specifiers **never** result in a number in scientific notation. In order to preserve as much of the number as possible, commas are not included in the format used by the FirebaseSQL database. At minimum, the `GetDouble` method of the FbDataReader should be able to convert numbers stored as text in this form back to a double value, which it does because it uses the `InvariantCulture`.

**Comment by BFuerchau** [comment]
You may have no probem, when you set at runtime the CurrentCulture to InvariantCulture, because you have also the CurrentUICulture to show data to the client.

The standard casts of _database_ i know, uses in most cases the currentculture of the system, where the server runs. This may be another than the client use. Normally the cast of datatime accepts currentculture and ISO which is the InvariantCulture. The Timestamp from SQL-Server contains an additional "T" between date and time.
If i read DateTime from the SQL-Sever i must often use the SQL-Servers own convert-function, to get an iso string instead of the stored datetime.

And when you tell the framework, that a text column is another type than string, than is this a bad design. Make your own wrapper for converting. The main problem is the communication between the client and the server. Parameters are serialized automatically with the typed GetValue, also the GetDate()/GetTime() and this values are converted to an integervalue. which is than invariant.
Did you have a look at the database, which value is really stored in the textcolumn?

If you dont accept this driver as is, you can download a copy, change them und use this. You can also create a nuget package for your application.
As you can see, the ticket is closed, the problem would not be solved, what i understand, because it is not an error.
If a database type doesnt match the parameter/fieldtype, the result may by not usable.

Have you looked at:
https://dev.to/adrianbailador/automapper-mastering-complex-mappings-in-c-46am

**Comment by deAtog** [comment]
@BFuerchau What you're overlooking here is that the user is not passing strings to the the database to store into the text column. As stated in the original issue, they are inserting a DateTime **object**. That **object** is ultimately stored as text by the database due to the column type. If they were inserting a string and retrieving a string, then the format and the parsing of that string is up to them. The following **are not** the same:

1. Inserting a DateTime **object** value using a parameterized query with a DateTime parameter type into a text column.
    - The conversion to text happens at the database layer. The FirebirdSQL client driver converts the DateTime **object** to a timestamp and then the server converts it to text using the [datetime_to_text](https://github.com/FirebirdSQL/firebird/blob/4dd80eda5d4bf1f41c15fa4795567896b06da39e/src/common/cvt.cpp#L2363) function and inserts it into the database field.
```c#
// For the purposes of this example the table must be initially empty.
// Insert a DateTime object directly into the database.
DateTime dtNow = DateTime.Now;
var sqlInsert = "INSERT INTO SomeTable (textfield) VALUES (@dtParam)";
using (var cmd = new FbCommand(sqlInsert, conn)) {
    var dtParam = cmd.CreateDbParameter();
    dtParam.ParameterName = "@dtParam";

    // Note: The parameter type is a DateTime and the value is a DateTime!
    dtParam.DbType = DbType.DateTime;
    dtParam.Value = dtNow;
    
    cmd.Parameters.Add(dtParam);
    cmd.ExecuteNonQuery();
}

// Query the DateTime value from the database. This currently fails if the current culture is not en-US!!
var sqlSelect = "SELECT textfield FROM SomeTable";
using (var cmd = new FbCommand(sqlSelect, conn))
using (var reader = cmd.ExecuteReader()) {
    // Read the first and only result in the table.
    reader.Read();

    DateTime dtThen = reader.GetDateTime(reader.GetOrdinal("textfield"));
}
```
2. Inserting a DateTime **string** value using a parameterized query with a **string** parameter type into a text column.
    - The user converts the DateTime **object** to a string in whatever format they want. The FirbirdSQL client driver sends the string to the database which inserts it into the database field unmodified.
```c#
// For the purposes of this example the table must be initially empty.
// Insert a DateTime as a specially formatted string into the database.
DateTime dtNow = DateTime.Now;
var sqlInsert = "INSERT INTO SomeTable (textfield) VALUES (@dtParam)";
using (var cmd = new FbCommand(sql, conn)) {
    var dtParam=  cmd.CreateDbParameter();
    dtParam.ParameterName = "@dtParam";

    // Note: The parameter type is a String and the value is a String!
    dtParam.DbType = DbType.String;
    dtParam.Value = dtNow.ToString("yyyy'/'MM'/'dd HH':'mm':'ss.ffffff", CultureInfo.InvariantCulture);

    cmd.Parameters.Add(dtParam);
    cmd.ExecuteNonQuery();
}

var sqlSelect = "SELECT textfield FROM SomeTable";
using (var cmd = new FbCommand(sqlSelect, conn))
using (var reader = cmd.ExecuteReader()) {
     // Read the first and only result in the table.
     reader.Read();

     // The value was inserted using a custom formatted string.
     // The value must be read as a string and parsed manually!
     string sThen = reader.GetString(reader.GetOrdinal("textfield"));

     DateTime dtThen = DateTime.ParseExact(sThen, "yyyy'/'MM'/'dd HH':'mm':'ss.ffffff", CultureInfo.InvariantCulture);
}
```

In the first case, the user should be able to call FbDataReader.GetDateTime() to retrieve the original DateTime **object** value that was stored in the database. In the second case, the user would be expected to call FbDataReader.GetString() and parse the value they  originally stored. **FbDataReader.GetDateTime is not expected to correctly parse data from case 2 above.** In other words, if a user formats an DateTime, integer, double, or whatever else into a string and inserts it into the database using a **string parameter** it is their responsibility to parse the string value from the database back into the original object. If they insert the value using the appropriate DbType, then they should be able to retrieve the original value using the Get* methods of the FbDataReader. **FbDataReader.GetDateTime is the only method that fails to do this properly.**

**Comment by BFuerchau** [comment]
All is clear, but as you can see in the answer from the developer, this can work only, if all clients use the same cultureinfo.
It is with all databases the same thing. This happens also with strings, when you doesnt hase the same codepage between clients and the column doesn't support unicode. The dateformat is the same thing.

Make your own copy and everything works.

**Comment by deAtog** [comment]
> All is clear, but as you can see in the answer from the developer, this can work only, if all clients use the same cultureinfo.

That's the problem. **It DOES NOT work even if all of your clients have the same CultureInfo.** The date format used by FirebirdSQL **IS NOT** determined by the current culture of the client or the database. Go read the [datetime_to_text](https://github.com/FirebirdSQL/firebird/blob/4dd80eda5d4bf1f41c15fa4795567896b06da39e/src/common/cvt.cpp#L2363) function.

If all clients are using a culture of "ar-SA", case 1 above fails with a FormatException when calling `FbDataReader.GetDateTime()`. Case 1 currently only works if the DateTime format of the current culture matches that of "en-US", which is by definition the same as the `InvariantCulture`. An application calling `FbDataReader.GetDateTime()` should not have to switch the current culture of the thread or application to the `InvariantCulture` just to make `FbDataReader.GetDateTime()` work correctly.

Code pages are irrelevant here as code pages are only used for ASCII extended characters. All characters used by FirebirdSQL for converting a timestamp to text are part of the standard ASCII characters with ascii codes in the range of 0-127. There is no need to get into Unicode or other character formats here as they are irrelevant to the issue.

**Comment by cincuranet** [comment]
Looking at your code, I'm not sure I originally understood your issue correctly. This code (I had to fix it, because you used couple of wrong methods)

```csharp
CultureInfo.CurrentCulture = CultureInfo.GetCultureInfo("ar-SA");

const string ConnectionString = "database=localhost:some-database;user=sysdba;password=masterkey";
FbConnection.CreateDatabase(ConnectionString, overwrite: true);
using var conn = new FbConnection(ConnectionString);
conn.Open();
using (var cmd = new FbCommand("RECREATE TABLE SomeTable (textfield VARCHAR(255))", conn))
{
	cmd.ExecuteNonQuery();
}
// For the purposes of this example the table must be initially empty.
// Insert a DateTime object directly into the database.
DateTime dtNow = DateTime.Now;
var sqlInsert = "INSERT INTO SomeTable (textfield) VALUES (@dtParam)";
using (var cmd = new FbCommand(sqlInsert, conn))
{
	var dtParam = cmd.CreateParameter();
	dtParam.ParameterName = "@dtParam";

	// Note: The parameter type is a DateTime and the value is a DateTime!
	dtParam.DbType = DbType.DateTime;
	dtParam.Value = dtNow;

	cmd.Parameters.Add(dtParam);
	cmd.ExecuteNonQuery();
}

// Query the DateTime value from the database. This currently fails if the current culture is not en-US!!
var sqlSelect = "SELECT textfield FROM SomeTable";
using (var cmd = new FbCommand(sqlSelect, conn))
using (var reader = cmd.ExecuteReader())
{
	// Read the first and only result in the table.
	reader.Read();

	DateTime dtThen = reader.GetDateTime(reader.GetOrdinal("textfield"));
}
```

round-trips the value fine. The `datetime_to_text` is irrelevant here. Also, your _are inserting a DateTime object. That object is ultimately stored as text by the database due to the column type_ is not correct. The conversion happens client-side, not server-side.

**Comment by BFuerchau** [comment]
The opinion is, when the datatime is stored with en-US (same as invariant) per default in text columns, the reader.GetDateTime sould also use InvariantCulture instead of CurrentCulture.
The same is also in case of double, where the server stores invariant, the reader.GetDouble also use InvariantCulture.
These seams to be inconsistent or inconsequent.
My opinion is, when i store information in a database not in matching database types, than this is my problem and not the problem of the databasemanager or database driver. I have such problems with each database.
I don't understand, why the user uses text instead of timestamp column for a DateTime. Than the problem is gone.

**Comment by deAtog** [comment]
@cincuranet The original issue was observed by switching locales from "en-US" to "ar-SA" in between the insert and the select. This issue exists and is problematic because the behavior changes based on the current culture.

In my original bug report, I noted that I did not know how FirebirdSQL converted DateTime to text. I attempted to analyze the code manually, but I failed to understand it correctly. My original understanding was that the DateTime was being converted to a Timestamp and then sent to the server as a Timestamp. What I missed was that every statement with parameters is prepared before execution and the parameters of the prepared statement are filled according to the field type that the server expects rather than the type of parameter specified by the client. As such, attempting to insert a DateTime value into a VarChar column results in the DateTime value being converted to text by the client and not the server.

I apologize for not fully understanding the issue. I propose the following solutions:

### Solution 1 - User specified culture
1. Add a connection string parameter to specify the culture used for conversions.
    - This can default to CultureInfo.CurrentCulture to maintain backwards compatibility.
    - It can then be changed to CultureInfo.InvariantCulture in a major release as a breaking change so that the behavior is more consistent by default.
2. Replace all instances of CultureInfo.CurrentCulture with the connection specific culture.
3. Find and correct any conversions using ToString() to call Convert.ToString with the connection specific culture.

### Solution 2 - User specified conversion method
1. Add a connection string parameter that specifies the behavior used for conversions.
    - Conversions are performed using the current culture as they are now.
    - Conversions are performed according to how FirebirdSQL performs casts.
2. Replace all instances of CultureInfo.CurrentCulture and ToString() with conversions based on the above parameter.

### Solution 3 - User specified compliance mode
1. Add a connection string parameter that determines how the client transmits parameter values for statements and parses values retrieved from the server.
    - The client converts parameter values according to the type returned from preparing the statement. (non-standard mode)
    - The client sends the parameter value to the server matching the closest database type that matches the client provided parameter type. For example, a DateTime parameter would always be converted and sent as Timestamp value. (standard mode.)
2. Values retrieved from the server when in standard mode would be expected to be converted according to the standard conversion rules used by FirebirdSQL.

    **_Note:_** This may only be possible if FirebirdSQL allows the client to explicitly set the parameter type according to the value provided.

---

### Other Considerations
If FirebirdSQL supports executing statements with parameters without preparing them first, the client should honor this mode of execution for the following reasons:
1. The DbCommand object has a Prepare() method that is used to explicitly prepare statements prior to execution.
2. Preparing a statement requires an additional round trip to the server to analyze the query and create an execution plan.
3. Prepared statements are slower if the statement is only executed once.

---

If I had a choice of which of these solutions to implement, I would likely choose Solution 3 if possible as it more closely aligns with behavior by other databases and would match the expected behavior of executing an unprepared statement.

**Comment by BFuerchau** [comment]
As i told above, the problem is, you use the wrong parameter type to the column type.
If you want to store anything in a text/string column than you are responsable for the correct database.
When you analyse the way, how parameters are sent to the server, than you are right, that all values are converted to string with invariant culture except the datetime.
Because of the various destination type, the value must have a date and time separatly. So these values are sent as DataTime in 2 Bigint-Values in a unified way, regardless of the culture. So the server gets a timestamp and must write this than in the text column. In this case, the server uses the invariant datatime format, which is en-us in mm/dd/yyyy hh.mm..ss pp (pp = am/fm).
So you have no chance in the driver, to do conversions.

What i not understand is, why you don't use a timestamp type in the column? If you do so, then the problem disapears.
When i want to store an unified timestamp in a text column, i format this value always to ISO-Format yyyy-MM-dd HH:mm.ss. The main point here is also, that this value can used by "order by" or filtered with greater, less, between and also date calculations.
The invariant date string is useless for all this database features.

And i think, a breaking change is also useless.

By the way:
**_Each database must prepare a sql string for execution!_**

**Comment by deAtog** [comment]
@BFuerchau As I said before, I don't disagree that users shouldn't be storing DateTime values in a text field. However, I disagree with your assessment that if someone does do so that the client driver should fail unexpectedly due to the current culture of the machine. 

If Solution 3 isn't a possibility, then maybe the best option would be to implement Solution 2 in combination with Solution 1. In standard mode, inserting a DateTime value into a text field would convert it to the ISO-Format that FirebirdSQL currently uses. In non-standard mode, it would convert it based on either the current culture or the user specified culture in the connection string. This would allow data to be insert/read in a consistent manner for existing databases. This would not be a breaking change unless standard mode was made the default mode.

> Because of the various destination type, the value must have a date and time separatly. So these values are sent as DataTime in 2 Bigint-Values in a unified way, regardless of the culture. So the server gets a timestamp and must write this than in the text column. In this case, the server uses the invariant datatime format, which is en-us in mm/dd/yyyy hh.mm..ss pp (pp = am/fm).
So you have no chance in the driver, to do conversions.

If the server received a timestamp value, I would expect it to convert it to text using the datetime_to_text function that I linked above. The format used is the standard ISO-Format with fractional seconds out to four decimal places.

**Comment by BFuerchau** [comment]
I'm programming databases now since 50 years and it is every time the responsible of the developer, to store and retrieve the correct data. The datbase manager works with typed columns very fine.
You can store in textfields any value you prefer, but it is not the database or driver, if you dount use the text columns type.
It is everyones knowledge, if you store not text values in text columns, that the database may have unexpected results.

The prepare is nessecary, because parameters must match the typ where it is stored or with wich expression it must be compared.
As i told you, if you want another driver, take the source and create your own.

It could not be, that modifications at a standard modul should solve only 1 mistake of 1 developer.

I think, this thread can be finally close;-).

**Comment by deAtog** [comment]
@BFuerchau I have been around as long as desktop computers have. It is standard practice for a database to either convert incoming data into the destination field type in a standardized way or to throw an error upon receiving incompatible data. FirebirdSQL's cast from a Timestamp to a character field is well defined, but this client driver is not following that standard and has thus created an issue where inserting/retrieving a DateTime into/from a character field causes inconsistent behavior. The proposal here is to correct this defect so that the data is stored in a more consistent format and to maintain backwards compatibility with current databases.

The DataReader in .Net is an abstraction layer on top of the data. The Get* methods are expected to be able to perform implicit casts on data read from the database. As we have discovered, these methods do not work reliably because of how data is insert and because of how some of these methods are implemented. It's easy to try and chalk this up to user error, but if that were the case the driver shouldn't have allowed the data to be insert in the first place.

I think the solution I have proposed above will lead to a better user experience and better reliability for the FirebirdSQL users. This is not an attempt to correct a single developer's mistakes, it's an attempt to prevent them from occurring to begin with.

---

I'm not that familiar with the FirebirdSQL protocol, so I don't know if it supports executing a statement without preparing it first. I found a C API function called isc_dsql_execute_immediate that appears to be capable of executing a query with a list of provided parameters, but there is not much documentation on it and even less on how it is implemented. Avoiding a round trip here would greatly improve performance if this feature is available.

**Comment by BFuerchau** [comment]
The prepare isn't a function of protocol, it is a function of the database manager.
The driver didn't analyse, what you send to the server. He didn't know, if the parameter type matches the requirement, he send it as is to the server.
The Server than prepares the commandtext, analyse syntax, type of select/insert/update/..... check the parameters from ? to the corresponding parameter.
If this is all done, plan must be check, if the sql isn't cached.
Than finally the execute of the command happens and perhaps you get an resultset.

So prepare isn't a function of the driver, it is a central und urgent function of the database manager.

The reader could cast to any what you select, but to your own risk. You can do also cast in SQL, with your own risk.

You didn't answered the question, why you can't use date oder timestamp for datetime fields? This is the main problem.
You can't use a Textfield for the many datetime-calculations, you can not filter greater/less or sort.
But its your choise;-).

**Comment by deAtog** [comment]
@BFuerchau A user can use a timestamp column for storing a datetime that is not an issue. The issue is if a database administrator creates a text field for storing a datetime while the developer thinks it is a timestamp column and treats it as such. In this case, insert and selects work so long as the current culture remains the same. Only when the culture changes is the problem discovered, and fixing the issue becomes problematic. Allowing the user to specify a specific culture to be used for conversions somewhat mitigates this issue, but avoiding the use of culture all together completely eliminates it. That is why my proposed solution to this issue is to provide a standard mode and a non-standard mode where the user can specify the culture used in the non-standard mode.

**Comment by BFuerchau** [comment]
I have checked the behaviour on my german system, with german language.
If i insert a timestamp in a table with a textfield, the database manager store the value as ISO, e.g. 2026-05-12 11:07:55.4820.
If in your environment the database manager stored the datatime as Invariant, may be it seams to be invariant. The language from the firebird server is set to en-us (MM/dd/yyyy...), so your datetime is stored as this.
When you now read the datetime on the client side in another culture as the servers culture you must fail in case of conversion.
You can change your query to retrieve the textfield as "cast(text as timestamp)", which then fails, if the content in't a local server culture timestamp.

All your solutions won't work!
1.  a culture doesnt work, because the conversion is done on the server after the prepare is done and only than, if the parameter type doesn't match the required type. Most DBM's, i know, in case of query or calculations try to cast the column or expression to the parameter type, which may than fail. The parameter is always sent as type you specify, any client culture isn't relevant and nothing would be converted.

2. How should this work, to specify a _method_ in connection_string_? As i told in 1, the decision, if a conversion must happen, depends on the context. How shold _the server_ call the method? For this, you can write a before trigger, to perhap correct the content. And should this method used for all datetime parameters and results? I don't think so!

3. The prepare is _always done_ at the server side, also if you don't call them explicit. If you define parameters, that wouldn't change. If you don't define parameters, the parameters are created automatically, depending on the context where a parameter is used. So in most cases for your problem you get a string parameter. If you store than values, you get perhaps a type mismatch.

The main mistake is here: 
"The issue is if a database administrator creates a text field for storing a datetime while the developer thinks it is a timestamp column "
This has to be corrected.
**An administrator should not create tables with wrong column types if these doesn't match the requirements!**
If so, the client have to work with the same culture (Region) as the server, to get no errors.

You will also have later problems with third party software like Power BI or similar tools. The column is a text column and never automatic recognized as datetime, so you have to convert it with additional overhead. 
If you need ony a part of your data for analysis, the filter for text works not when the datetime is stored as en-us MM/dd/yyyy.

**Comment by deAtog** [comment]
@BFuerchau The FirebirdSQL datetime text format is not en-US. That format used is well defined in the FirebirdSQL source code as I described above. In standard mode, the client driver would always store datetime values into character fields in that form to match what the server does. This is an issue in the client driver because:
1. The client sends the DateTime value as text instead of as a timestamp (according to [Row BLR (Binary Language Representation)](https://firebirdsql.org/file/documentation/html/en/firebirddocs/wireprotocol/firebird-wire-protocol.html#wireprotocol-row-data-blr)  the client may send types that are not what the server expects, but the types must be convertible).
2. The client currently converts non-string objects to strings by calling ToString() which defaults to using the client's current culture for string conversions. This format does not match the form used by the FirebirdSQL datatbase. The bulk of the conversions happens in [GdsStatement.WriteRawParameter](https://github.com/FirebirdSQL/NETProvider/blob/a61996cb5734af1e90a4083e9c9933e4de8b2e05/src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsStatement.cs#L1225) with some modifications being made during [FbCommand.UpdateParameterValues](https://github.com/FirebirdSQL/NETProvider/blob/a61996cb5734af1e90a4083e9c9933e4de8b2e05/src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbCommand.cs#L1111) for Binary, Text, Array, and Guid destination field types. Once the statement is prepared, a DbDataType of Char or VarChar will cause any non-string values to be converted to strings by the Object.ToString() method.
---
>1. a culture doesnt work, because the conversion is done on the server after the prepare is done and only than, if the parameter type doesn't match the required type. Most DBM's, i know, in case of query or calculations try to cast the column or expression to the parameter type, which may than fail. The parameter is always sent as type you specify, any client culture isn't relevant and nothing would be converted.

The client doesn't currently work this way. The client currently receives a set of field descriptors from the server after preparing the parameterized statement. The client then writes the BLR used for the execute statement based upon the field types of the descriptors. A DbType.DateTime parameter being insert into a DbDataType.VarChar field is currently converted to a string by the client using the current culture. Providing a culture to use for this conversion and for other conversions would partially solve the conversion issue.

> 2. How should this work, to specify a method in connection_string_? As i told in 1, the decision, if a conversion must happen, depends on the context. How should the server call the method? For this, you can write a before trigger, to perhap correct the content. And should this method used for all datetime parameters and results? I don't think so!

Obviously, the user of the client driver can't provide a conversion method. They could however specify how conversions should happen in the client driver. In non-standard mode, data conversions in the client driver would use either the current culture or the culture specified in the connection string. In standard mode, conversions would either be offloaded to the server (recommended), or performed in the client as they would be in the server.

> 3. The prepare is always done at the server side, also if you don't call them explicit. If you define parameters, that wouldn't change. If you don't define parameters, the parameters are created automatically, depending on the context where a parameter is used. So in most cases for your problem you get a string parameter. If you store than values, you get perhaps a type mismatch.

Correct, the server always prepares a statement before it is executed. However, most databases allow you to skip statement preparation from the client prior to execution. In most cases, the client never has any knowledge of what the destination field types are and all conversions are handled by the server. FirebirdSQL however is not like other databases as it tells the client what types the server expects for the fields. This driver currently sends parameter types and values according to the field descriptors returned from preparing the statement. This differs from how most other database implementations work. The BLR documentation states that the client may send different BLR than what the server expects so long as the types are convertible.

The BLR documentation also hints at the existence of a op_exec_immediate2 instruction which accepts parameters. This op code is currently undocumented, but the wire protocol probably wouldn't be too difficult to reverse engineer or brute force. I suspect this operation allows executing a statement without calling prepare first. In this case, the BLR sent would be generated from the client specified parameter types.

**Comment by BFuerchau** [comment]
From what you get your knowledge?
The main class for communication is
XdrReaderWriter
Each type has a specific  serialize/deserialize function, which converts type to/from byte-Array.
You never find a ToString()-call.

Prepare on the server can't be supressed. This is technologie depended, for all databases. What prepare does and why it is needed, i have already told.
How else can the server decide, what you want, which parameter maps to field or expression?
The firebird server tells me the parameter types, when you explict prepare without any own parameter, this is also standard for most databases.
The driver handles the comandtext independet from the parameters. It makes no sence, that the driver should make a syntaxcheck and retrieve table informations to handle a type  conversion for parameters. This is a normal server task.
The firebind client writes the parameters with only nessesary conversions for serializing values. The driver doesn't know any over the usage of parameters on the server. So conversions are done so, that the server get the same value on deserialization.
Any individual function or format will break the protocoll.
What you can do here is, define all parameters as string and look what happens. The server must than interpret, how strings are than back converted to corresponding types.

May be you have worked in the past with SQLite. This database works without types and stores any value in any column.
Than it is your responsibility, how to process with data.
But Firebird is strongtyped!

Than you didn't looked at the source of the driver, how parameters are sent.
Finally all parameters are converted with the TypeEncoder class.
For Date and Time, you will find the encoder, which calculates int, which is sent to the server.

`	static int EncodeDateImpl(int year, int month, int day)
	{
		if (month > 2)
		{
			month -= 3;
		}
		else
		{
			month += 9;
			year -= 1;
		}

		var c = year / 100;
		var ya = year - 100 * c;

		return ((146097 * c) / 4 + (1461 * ya) / 4 + (153 * month + 2) / 5 + day + 1721119 - 2400001);
	}

	public static int EncodeTime(TimeSpan t)
	{
		return (int)(t.Ticks / 1000L);
	}
`
Don't ask me, why.

"Correct, the server always prepares a statement before it is executed. However, most databases allow you to skip statement preparation from the client prior to execution. In most cases, the client never has any knowledge of what the destination field types are and all conversions are handled by the server. "

Sorry, but this can only say, who nothing knows over database programming.
The skip prepare means, that you don't need an explicit prepare before execute. 
The server must prepare each statement before first execute.
After the first execute, the statement is prepared and cached.

May be, the client does not know the types, but the _developer should know!_
If you don't know the structure of tables and types of a destination field, all automatically conversion can be failed.

When you use frameworks like EF or strongtyped datatable objects, like supported from VS, you have automatically classes, with properties for each field in the corresponding type of the server.
So in communication between program and database all works fine, because the client knows everything about types.
You must know, that a string can not be longer than the defined size of the field, or a fixed decimal as 10 digits with 2 fractions.

And thats your problem. I don't understand why you not let alter the table to the correct fieldtype and all problems are gone, without any changes to the driver, database or your program.

The execute immediate is an operation which comes from languages with embedded sql. There you have often explicit language statements like
`
commandString = "Select * from mytable where key = ? and status = ?";
exec sql prepare mystmt :commandString;
exec sql call mystmt using :p1, :p2;

or

exec sql execute immediate commandString using :p1, :p2;
`
With the modern languages you don't need it anymore. But the serverside prepare must always happen.

**Comment by deAtog** [comment]
@BFuerchau My knowledge comes from reading and testing the source code of this client driver. Here is all the information I know about the [FirebirdSQL Wire Protocol](https://firebirdsql.org/file/documentation/html/en/firebirddocs/wireprotocol/firebird-wire-protocol.html).

> The main class for communication is
XdrReaderWriter
Each type has a specific serialize/deserialize function, which converts type to/from byte-Array.
You never find a ToString()-call.

Here a trace of what happens when you call `FbCommand.ExecuteReader`
- [FbCommand.ExecuteReader](https://github.com/FirebirdSQL/NETProvider/blob/34acfda601380e4626605191def614c2fc9e8f34/src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbCommand.cs#L582)
  - [FbCommand.ExecuteCommand](https://github.com/FirebirdSQL/NETProvider/blob/34acfda601380e4626605191def614c2fc9e8f34/src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbCommand.cs#L1446)
      - [FbCommand.Prepare](https://github.com/FirebirdSQL/NETProvider/blob/34acfda601380e4626605191def614c2fc9e8f34/src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbCommand.cs#L1307) 
          - Creates and prepares a statement by sending the SQL to the server.
          - Retrieves the field and parameter descriptors.
          - Upon success, sets FbCommand._statement to the prepared statement.
      - _statement.Execute -> [GdsStatement.Execute](https://github.com/FirebirdSQL/NETProvider/blob/34acfda601380e4626605191def614c2fc9e8f34/src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsStatement.cs#L300)
          - [GdsStatement.SendExecuteToBuffer](https://github.com/FirebirdSQL/NETProvider/blob/34acfda601380e4626605191def614c2fc9e8f34/src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsStatement.cs#L766)
              - [GdsStatement.GetParameterData](https://github.com/FirebirdSQL/NETProvider/blob/34acfda601380e4626605191def614c2fc9e8f34/src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsStatement.cs#L1804)
                  -  IDescriptorFiller.Fill -> [FbCommand.Fill](https://github.com/FirebirdSQL/NETProvider/blob/34acfda601380e4626605191def614c2fc9e8f34/src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbCommand.cs#L1110)
                      - [FbCommand.UpdateParameterValues](https://github.com/FirebirdSQL/NETProvider/blob/34acfda601380e4626605191def614c2fc9e8f34/src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbCommand.cs#L1111)
                          - This updates the parameter descriptors of the statement with the values of the FbCommand parameters.
                          - The VarChar descriptor's DbValue.Value is set to the DateTime object passed as the parameter value.
                  - [GdsStatement.WriteParameters](https://github.com/FirebirdSQL/NETProvider/blob/34acfda601380e4626605191def614c2fc9e8f34/src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsStatement.cs#L1174)
                      - [GdsStatement.WriteRawParameter](https://github.com/FirebirdSQL/NETProvider/blob/34acfda601380e4626605191def614c2fc9e8f34/src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsStatement.cs#L1225) - field.DbDataType is DbDataType.VarChar
                          - [DbValue.GetString](https://github.com/FirebirdSQL/NETProvider/blob/34acfda601380e4626605191def614c2fc9e8f34/src/FirebirdSql.Data.FirebirdClient/Common/DbValue.cs#L150)
                              - [Object.ToString](https://github.com/FirebirdSQL/NETProvider/blob/34acfda601380e4626605191def614c2fc9e8f34/src/FirebirdSql.Data.FirebirdClient/Common/DbValue.cs#L161) - DbValue.Value is converted to a string based on the current culture.
              - _parameters.ToBlr -> [Descriptor.ToBlr](https://github.com/FirebirdSQL/NETProvider/blob/34acfda601380e4626605191def614c2fc9e8f34/src/FirebirdSql.Data.FirebirdClient/Common/Descriptor.cs#L112)

As you can see above, the client fills the parameter descriptors from the server using the client values. It then generates the BLR of those descriptors. It then sends the BLR along with the parameter data when executing the statement. The parameter types sent to the server are the same as those received when preparing the statement. As such, the parameter is sent as a VarChar after the DateTime value was converted to a string by the client.

Because of the way this client operates, the following test using a double fails.

```C#
	[Test]
	public async Task InsertDoubleIntoVarChar()
	{
		var r = new Random();
		double d = r.NextDouble() * 1e9;
		var culture = CultureInfo.CurrentCulture;

		try
		{
			CultureInfo.CurrentCulture = new CultureInfo("de-DE", false);

			await using (var command = new FbCommand("insert into TEST (int_field, varchar_field) values (1234, @d)", Connection))
			{
				var param = command.CreateParameter();

				param.DbType = DbType.Double;
				param.Value = d;
				param.ParameterName = "@d";

				command.Parameters.Add(param);

				var ra = await command.ExecuteNonQueryAsync();
			}

			await using (var command = new FbCommand("select varchar_field from TEST where int_field = 1234", Connection))
			await using (var reader = await command.ExecuteReaderAsync())
			{
				reader.Read();

				var j = reader.GetDouble(0);

				Assert.AreEqual(d, j);
			}
		}
		finally
		{
			CultureInfo.CurrentCulture = culture;
		}
	}
```
This test fails is due to the same issue as I originally reported, just using a different type. In this case, the test fails because the ToString() call in the DbValue.GetString() method returns a value with a comma as the decimal separator and FbDataReader.GetDouble() parses the string retrieved using the InvariantCulture. No exception is thrown, because the InvariantCulture ignores the comma but returns a number magnitudes larger than the original.

In my opinion, the client should ignore the parameter descriptors received and generate them directly from the parameters of the FbCommand. This would avoid converting from the parameter type to the descriptor type in the client. Thus a Double or DateTime value would always be sent to the server as a Double or a Timestamp instead of as whatever parameter descriptor type the prepare statement returns. If the server is not able to perform the conversion, then the execution fails as expected. This is also consistent with how one would execute a statement without preparing it first if implemented.

### Notes
While stepping through and testing the above call sequence, I noticed some potentially error prone code in the DbValue.GetString implementation. As implemented, if the field type is DbDataType.Text and the stored value is a long then the value of the DbValue is converted to a string using DbValue.GetClobData which assumes that there is a statement associated with the DbValue. This could lead to a NullReferenceException when the DbValue is not associated with a statement. It could also result in a long value being incorrectly converted to a string when inserting into a Text column. 

_Update_: While this could be a problem, actual tests don't get that far. An invalid cast exception occurs in the `FbCommand.UpdateParameterValues` method due to trying cast a 64bit long value to a string. The cast here causes the client driver to crash if you attempt to insert any data type other than a string or a byte array into a BLOB SUBTYPE TEXT field.

**Comment by BFuerchau** [comment]
Ok, sorry you are right. 
The internal prepare retrieves the parameters from the firebird, with additional roundtrip, what other databases don't do.
This seams done to support named parameters, what originally isn't supported by the firebird itself.

Than merge existing parameters from the command to the retrieved parameters from statement.
This explanes also other misbehavior, that you get no error, when you execute a query or nonquery and forgot to create the needed parameters.
In this case, the retrieved parameters get there defaults or dbNull, but it should raise an error of missing parameters.

But as i told, when both, server and client uses the same CurrentCulture, you have no problem.
So my Insert with parameter inserts a date in german format "tt.MM.yyyy".
If you make an "insert into mytable values(current_tmestamp)", this is written in the field in ISO-Format "yyyy-MM-dd".

The  GetDateTime Supports both, CurrentCulture and ISO.
But the problem starts than with the order by, also indexes can't by used, which degrades performance

_Your explains confirm, that it is urgent to use the correct column type for the requirement data,_ because all others get problems in further processing.
I don't understand, that you will create work arounds with a lot of other problems and not correct the real cause, which solves all your problems. If you will use any value in text columns, you should convert it by yourself.

**Comment by deAtog** [comment]
@BFuerchau In my opinion, there are numerous problems with the current approach. Attempting to convert the client provided values to the server expected types in the client driver is the wrong approach in my opinion. This leads to unexpected behavior as demonstrated by the many examples I have provided. The expected behavior would be for the server to receive the same DateTime value or timestamp regardless of the client culture, and for the server to then convert it if required or return an error. The behavior of the server in this case is well defined and results in data stored in a consistent manner. This issue isn't related to just object to string conversions. String to numeric conversions have issues as well that result in a FormatException being thrown from the client rather than an FbException from the result of a server based conversion error.

The BLR specification states that the client is permitted to send the data however it wants. In my opinion that's what the client should do. Then the client only needs to coerce the parameter value to the client provided parameter type and let the server handle any other conversions. For a DateTime value, that means coercing the value to a timestamp and sending the value as a timestamp. For floats, double, long, int, decimal, the client should merely send those values to the server.

> I don't understand, that you will create work arounds with a lot of other problems and not correct the real cause, which solves all your problems. If you will use any value in text columns, you should convert it by yourself.

I understand you think this is a case of the user is doing it wrong. My issue with that opinion is that if that were the case, then the driver should throw a reasonable exception with an appropriate error message. E.g. Something along the lines of "Cannot insert a DateTime value into VarChar field {field}" with an exception type of FbException. This however would be a major breaking change with no advantages other than to mask the issues of this client driver.  The better approach is to address the behavior of the client and let the server deal with any data conversions that may be necessary.

**Comment by BFuerchau** [comment]
I did not have any problem with the driver, if i use the correct column type. And i think, all other developers, which uses these driver have also no problems. The driver exists since years. It seams, you are the only one. Sorry.

**Comment by deAtog** [comment]
@BFuerchau I'm probably the first one that noticed because I develop software that is meant for international use. I've run into these sort of issues before and am more aware of them. Many of these issues never crop up until the culture changes. The proper thing to do here is to write tests around every possible conversion that verifies expected functionality according to behavior of the server. If the server doesn't support a particular conversion, the client shouldn't either.

The solution I've recommended would correct these issues while maintaining backwards compatibility with existing databases.

**Comment by BFuerchau** [comment]
As i explained. your solutions arn't possible or has many works on client and server!
Also many other versions of drivers, Java, Python, ODBC or else, must do these also to stay compatible.

I'm using my solution also in multinational cultures, but i always use the correct type of column for the data.

Cause your wish for exception, i have checked that many database allows you to store any value in a textcolumn and convert between types, but throws errors if type isn't convertibale or the result doesn't match.
But it is always your responsible, to correct use the textcolumn in case of reading the value, so this is standard behavior.

As you have seen, that the _value of text column is a long, thats belongs to the protocoll of the firebird server. Long strings (i don't know at which lenght), arrays, clobs or binary aren't loaded in the fetch, so the _value contains a handle. To load the original value, you have to call the corresponding GetValue, GetString, or GetValues, to load the remaining data _before_ you can convert.
In case of other getters you get an error, long to DataTime, or the long value itself as int, double, and so on.

If any other developer works e.g. with frameworks like EF, Typed DataSets or TypedTables, your columns than are always text columns.

I don't think your wishes in this case becomes true.
To work _correct_ with each database manager (DB2, Oracle, SQL-Server and also Firebird), you must _use correct typed columns_ to avoid problems, special for multinational software and use of multinational clients in parallel.

**Comment by deAtog** [comment]
@BFuerchau The solutions I have suggested could all be completed from the client side. The amount of work varies from the method used. Here are the client-side changes required to make the client driver behave in a consistent manner:
1. Add culture to connection string for client side data conversions.
   - Difficulty: Easy
   - Impact on existing code: Low
2. Add operational mode parameter to connection string for client side driver for data conversions.
   a. Option 1 - Update client to send parameters based on client provided types.
      - Difficulty: Moderate
      - Impact on existing code: Moderate

   b. Option 2 - Update client to format and parse strings according to FirebirdSQL server conversions.
      - Difficulty: Low
      - Impact on existing code: Low

Just because other implementations have the same bugs, doesn't mean this isn't worth fixing.  Anything fixed here can always be ported to other drivers. Progress isn't made by accepting failure.

Anyone who uses this driver and accidentally uses a VarChar field for storing anything but text, currently would have a hard time migrating the data to a strictly typed column.  The built-in CAST function is designed to only handle conversions that the server performs. When operating in standards mode, these changes would align all conversions with those performed by the server allowing server side casts to correctly convert the data.

**Comment by BFuerchau** [comment]
You can do it by yourselve and make pushrequestes for this changes.
Be shure, that all modifications runs the tests, included by the project, complete errorfree.
Than perhaps your changes may be accepted.

In the past, a few years, i have done similar things concerning performance. My modifications allows more faster download of data from firebird. All tests run fine.
The answer: no one has any performance issues, so this suggestions aren't accepted and included.
So you can do the same way like i do: i load sometimes new versions of the project, reimplement my modifications so i have the fastes driver.
A few days before i have seen an issue, that someone needs long time to load 1 Million rows withe a single integer.
I have checked this with my version. Depending on other processes in my environment, my query of 1 single value give me 120.000 to 150.000 Rows per second! A Query with 55 Columns of mixed types give me 30.000 Rows per second.
I'm developing a data analysis application for web clients, i need performant reads from the database, which i get only with my own version of the driver.

**Comment by cincuranet** [comment]
This is interesting conversation, but I think it is going nowhere. Let me express where I stand on this.
 
Storing types in string/varchar is wrong (let alone when locales are involved) and sooner or later you'll experience issues (both server-side and client-side). The best way would be to reject the conversion in the provider and throw an exception. It is not like doing the conversion manually is difficult. But sooner or later people would ask for it anyway, with some reasonably looking reason. Anyway, that ship has already sailed.

That said, I'll open an issue and go through the codebase and make sure all the implicit conversions to/from string happen with invariant locale. It is going to be breaking change, but low priority in my eyes.

No settings, etc. Polluting connection string or some API for this already wrong approach is not worth it.

**Comment by deAtog** [comment]
@BFuerchau If I find time I will submit a pull request for changes. I've done a few tests locally already. Some tests have been more fruitful than others. 

One glaring issue I found yesterday that I overlooked was the implementation of the FbDataReader. All the Get* methods of the FbDataReader currently call the `FbDataReader.GetFieldValue<T>` method which then calls the type specific Get* method of the underlying DbValue. This leads to unexpected behavior.  The DbValue class should only return the correct FirebirdSql data type for the underlying field/column.  As you know, blobs, clobs, and arrays require special handling when reading the value. The initial value for these types is a 64bit int which is a handle to retrieve the actual value stored. Currently `FbDataReader.GetInt64` results in a call to `DbValue.GetInt64`, just as `FbDataReader.GetString` results in a call to `DbValue.GetString`. When a statement returns a clob, calling `FbDataReader.GetInt64` will return the handle used to retrieve the actual string value without actually reading the value of the clob. Likewise, calling `FbDataReader.GetString` on a statement that returns an Int64 will attempt to read a non-existent clob.

`FbDataReader.GetFieldValue<T>` should call the `FbDataReader.Get*` and all `FbDataReader.Get*` methods should merely call `DbValue.GetValue`, they should **not** call the `DbValue.Get*` methods except in a few instances like maybe `FbDataReader.GetStream`. Any conversions from the database specific types should occur in the FbDataReader, not in the DbValue class.

I still have reservations about the side effects of calling `DbValue.GetString` and other methods that deal with clobs, blobs, and arrays but I don't have a good idea for how to address them just yet. I have concerns there may be issues  in the client/server functionality if a query returns a blob, clob, or an array but the actual value is never retrieved. I have not checked if this causes a resource leak or other client/server issues.

**Comment by deAtog** [comment]
@cincuranet As stated above, I think all this can be done without causing breaking changes. The initial defaults for the connection parameters can be to preserve backwards compatibility and then be changed in a major release which breaks backward compatibility when the connection parameters are not explicitly set.

**Comment by BFuerchau** [comment]
@deAtog This behavior describes, what happens in case of incorrect type handling.
Correct your type in the table or use GetValues(objectArray), than you have a result and can do what ever you want.
The GetValue<T> are only helpers to get directly an unboxed value, depending on the stored Type, which also is incorrect, if the value is NULL or DbNull.
I work with many different databases and always i must handle the content depending of the columntype.
This is the reason to use always the GetValues() or load it in a DataTable-Object.
I don't understand why you are reluctant to change the table appropriately!

@cincuranet 
On the other hand, when i look to the driver, in case of absent of named parameters, you don't need a prepare before execute, because you can send the parameters as is in the defined types. Perhaps, the database throws than an error or may try convert automatically to the destination type, which may also throw errors. The database must prepare anyway.
The benefit is also, if you have more "?" in the commandtext as in the parameters, than i get an error. At the moment, you fill null or default, if a parameter is missing, thats also incorrect.
